sentry-android-core/src/main/java/io/sentry/android/core/SpanFrameMetricsCollector.java

```diff
@@ -68,63 +107,130 @@ public void onSpanFinished(final @NotNull ISpan span) {
     if (!enabled) {
       return;
     }
+
     if (span instanceof NoOpSpan) {
       return;
     }
+
     if (span instanceof NoOpTransaction) {
       return;
     }
 
-    @Nullable SentryFrameMetrics diff = null;
+    // ignore span if onSpanStarted was never called for it
+    final @NotNull String spanId = span.getSpanContext().getSpanId().toString();
     synchronized (lock) {
-      final @Nullable SentryFrameMetrics metricsAtStart =
-          metricsAtSpanStart.remove(span.getSpanContext().getSpanId());
-      if (metricsAtStart != null) {
-        diff = currentFrameMetrics.diffTo(metricsAtStart);
+      if (!spanStarts.containsKey(spanId)) {
+        return;
+      }
+    }
+
+    captureFrameMetrics(span);
+
+    synchronized (lock) {
+      if (runningSpans.isEmpty()) {
+        clear();
+      } else {
+        // otherwise only remove old/irrelevant frames
+        final SpanStart oldestSpan = runningSpans.first();
+        frames.headSet(oldestSpan.timeNanos).clear();
```

Comment:
i would remove the frames ended before the start of the first span, while this removes the frames ended before the end of the first span.
Otherwise concurrent spans could have some missing frames, correct?

Reply:
> i would remove the frames ended before the start of the first span
Yes, we do exactly that: Only frames which end (`frame.end`) before the oldest span start(`min(span.start)`) are removed here.

Reply:
ok, i've seen you updated it, and it's fine now